src/main/java/com/commercetools/sync/products/utils/ProductUpdateActionUtils.java

```diff
@@ -594,4 +610,49 @@ static AddVariant buildAddVariantUpdateActionFromDraft(@Nonnull final ProductVar
         @Nonnull final Supplier<List<T>> updateActionSupplier) {
         return executeSupplierIfPassesFilter(syncFilter, filter, updateActionSupplier, Collections::emptyList);
     }
+
+    /**
+     * Validate both old and new product have master variant with significant key.
+     *
+     * <p>If at least on of the master variants key not found - the error is reported to {@code syncOptions} and
+     * <b>true</b> is returned.
+     *
+     * @param oldProduct  old product to verify
+     * @param newProduct  new product to verify
+     * @param syncOptions {@link BaseSyncOptions#applyErrorCallback(String, Throwable) applyErrorCallback} holder
+     * @return <b>true</b> if at least one of the products have invalid (null/blank) master variant or key.
+     */
+    private static boolean haveInvalidMasterVariants(@Nonnull final Product oldProduct,
+                                                     @Nonnull final ProductDraft newProduct,
+                                                     @Nonnull final ProductSyncOptions syncOptions) {
+        boolean hasError = false;
+
+        final ProductVariant oldMasterVariant = oldProduct.getMasterData().getStaged().getMasterVariant();
+        if (isBlank(oldMasterVariant.getKey())) {
+            handleBuildVariantsUpdateActionsError(oldProduct, BLANK_OLD_MASTER_VARIANT_KEY, syncOptions);
+            hasError = true;
+        }
+
+        final ProductVariantDraft newMasterVariant = newProduct.getMasterVariant();
+        if (newMasterVariant == null || isBlank(newMasterVariant.getKey())) {
+            handleBuildVariantsUpdateActionsError(oldProduct, BLANK_NEW_MASTER_VARIANT_KEY, syncOptions);
+            hasError = true;
+        }
+
+        return hasError;
+    }
+
+    /**
+     * Apply error message to the {@code syncOptions}, reporting the product key and {@code reason}
+     *
+     * @param product     product which has sync error
+     * @param reason      reason to specify in the error message.
+     * @param syncOptions {@link BaseSyncOptions#applyErrorCallback(String, Throwable) applyErrorCallback} holder
+     */
+    private static void handleBuildVariantsUpdateActionsError(@Nonnull final Product product,
+                                                              @Nonnull final String reason,
+                                                              @Nonnull final ProductSyncOptions syncOptions) {
+        syncOptions.applyErrorCallback(format("Failed to build update actions on the product with key '%s'. Reason: %s",
```

Comment:
other update actions for the product will be built. Only variant-related update actions will be skipped. Please adjust error message accordingly.

Reply:
d519defc